plugins/print/plugin.js

```diff
@@ -6,41 +6,79 @@
-	lang: 'af,ar,az,bg,bn,bs,ca,cs,cy,da,de,de-ch,el,en,en-au,en-ca,en-gb,eo,es,es-mx,et,eu,fa,fi,fo,fr,fr-ca,gl,gu,he,hi,hr,hu,id,is,it,ja,ka,km,ko,ku,lt,lv,mk,mn,ms,nb,nl,no,oc,pl,pt,pt-br,ro,ru,si,sk,sl,sq,sr,sr-latn,sv,th,tr,tt,ug,uk,vi,zh,zh-cn', // %REMOVE_LINE_CORE%
-	// jscs:enable maximumLineLength
-	icons: 'print,', // %REMOVE_LINE_CORE%
-	hidpi: true, // %REMOVE_LINE_CORE%
-	init: function( editor ) {
-		// Print plugin isn't available in inline mode yet.
-		if ( editor.elementMode == CKEDITOR.ELEMENT_MODE_INLINE )
-			return;
-
-		var pluginName = 'print';
-
-		// Register the command.
-		editor.addCommand( pluginName, CKEDITOR.plugins.print );
-
-		// Register the toolbar button.
-		editor.ui.addButton && editor.ui.addButton( 'Print', {
-			label: editor.lang.print.toolbar,
-			command: pluginName,
-			toolbar: 'document,50'
-		} );
-	}
-} );
-
-CKEDITOR.plugins.print = {
-	exec: function( editor ) {
-		if ( CKEDITOR.env.gecko ) {
-			editor.window.$.print();
-		} else {
-			editor.document.$.execCommand( 'Print' );
+	CKEDITOR.plugins.add( 'print', {
+		requires: 'preview',
+		// jscs:disable maximumLineLength
+		lang: 'af,ar,az,bg,bn,bs,ca,cs,cy,da,de,de-ch,el,en,en-au,en-ca,en-gb,eo,es,es-mx,et,eu,fa,fi,fo,fr,fr-ca,gl,gu,he,hi,hr,hu,id,is,it,ja,ka,km,ko,ku,lt,lv,mk,mn,ms,nb,nl,no,oc,pl,pt,pt-br,ro,ru,si,sk,sl,sq,sr,sr-latn,sv,th,tr,tt,ug,uk,vi,zh,zh-cn', // %REMOVE_LINE_CORE%
+		// jscs:enable maximumLineLength
+		icons: 'print,', // %REMOVE_LINE_CORE%
+		hidpi: true, // %REMOVE_LINE_CORE%
+		init: function( editor ) {
+			var pluginName = 'print';
+
+			editor.addCommand( pluginName, CKEDITOR.plugins.print );
+			editor.ui.addButton && editor.ui.addButton( 'Print', {
+				label: editor.lang.print.toolbar,
+				command: pluginName,
+				toolbar: 'document,50'
+			} );
 		}
-	},
-	canUndo: false,
-	readOnly: 1,
-	modes: { wysiwyg: 1 }
-};
+	} );
+
+	/**
+	 * Allows to customise implementation of printing the editor's content,
+	 * provided by [Print](https://ckeditor.com/cke4/addon/print) plugin.
+	 * The provided value should be of {@link CKEDITOR.commandDefinition} type.
+	 *
+	 * ```javascript
+	 * // Using external API to generate PDF.
+	 * CKEDITOR.plugins.print = {
+	 * 	exec: function( editor ) {
+	 * 		var data = editor.getData();
+	 *
+	 * 		pdfAPI.generateFromHTML( data );
+	 * 	}
+	 * };
+	 * ```
+	 *
+	 * @singleton
+	 * @class CKEDITOR.plugins.print
+	 */
+	CKEDITOR.plugins.print = {
+		exec: function( editor ) {
+			var previewWindow = CKEDITOR.plugins.preview.createPreview( editor ),
+				nativePreviewWindow;
+
+			if ( !previewWindow ) {
+				return;
+			}
+
+			nativePreviewWindow = previewWindow.$;
+
+			// In several browsers (e.g. Safari or Chrome on Linux) print command
+			// seems to be blocking loading of the preview page. Because of that
+			// print must be performed after the document is complete.
+			if ( nativePreviewWindow.document.readyState === 'complete' ) {
+				return print();
+			}
```

Comment:
This code seems to produce the issue with missing images on Chrome. Removing it executes print command once images have been loaded. That's weird because according to docs `readyState === 'complate'` should be called once images are loaded, so it's seems like a bug of Linux Chrome. Nevertheless, it's also set **before** `load` event (see https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState) which invalidates fix for images.

If I remember correctly, `load` event is incorrectly propagated on Safari? Maybe we just need to print preview immediately on Safari? It seems to work correctly without waiting for load event.

Reply:
> If I remember correctly, `load` event is incorrectly propagated on Safari? Maybe we just need to print preview immediately on Safari?

It's not that simple… probably `load` propagates correctly, but in some cases, window can be opened _after_ `load` event. At the same moment not waiting for `load` event can cause invoking `print` command on unstyled content without images. In such case, checking `readyState` ensures that `print` is invoked after loading the document.

I'm not sure if we shouldn't just report bug in Chrome and leave it as is. Alternatively maybe adding `setTimeout` inside `readyState` check will do the trick 🤔

Reply:
So, maybe just use `readyState === 'complete'` only for Safari, for the rest `load` event?

Reply:
Ok, so it looks like disabling browser addons significantly reduce the issue and as we already checked this issue with another dev on Linux machine - I'm for keeping this implementation as it's clean and should work correctly. Any related issue is rather an edge case with system distribution than the implementation problem.